### ISSUE TITLE
Adapter restart when changing values for Eurotronic SPZ Thermostat

### ISSUE THREAD

**Issue by tsvsj** [ioBroker/ioBroker.zigbee#807]
Hi,

I've implemented my first Eurotronic SPZ Thermostat today. Pairing was working well, while changing any values to the Thermostat doesn't. I've tried to change several values, without any effect to the device. For now I'm not sure whether a signal strength between 25-35 is enough, but as I have other battery driven Zigbee devices in the same room, I assume, that the signal strength is quite enough.

![image](https://user-images.githubusercontent.com/39584211/95122935-7fa1fd00-0751-11eb-8a65-63f9c8ee3805.png)

However, while inspecting the reason, that the commands are not sent or received properly, I've seen a timeout issue with zigbee-herdsman, which results in a adapter restart, each time I'm trying to change a value.

```
2020-10-05 21:21:30.741  - error: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2020-10-05 21:21:30.743  - error: zigbee.0 (1450) unhandled promise rejection: Timeout - 6933 - 1 - 2 - 513 - 4 after 10000ms
2020-10-05 21:21:30.744  - error: zigbee.0 (1450) Error: Timeout - 6933 - 1 - 2 - 513 - 4 after 10000ms
    at Timeout._onTimeout (/opt/iobroker/node_modules/zigbee-herdsman/dist/utils/waitress.js:44:24)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
2020-10-05 21:21:30.746  - info: zigbee.0 (1450) cleaned everything up...
2020-10-05 21:21:30.749  - info: zigbee.0 (1450) Zigbee: disabling joining new devices.
2020-10-05 21:21:31.042  - info: zigbee.0 (1450) terminating
2020-10-05 21:21:31.044  - info: zigbee.0 (1450) Terminated (NO_ERROR): Without reason
2020-10-05 21:21:32.000  - info: host.xxxxxx instance system.adapter.zigbee.0 terminated with code 0 (NO_ERROR)
2020-10-05 21:21:32.002  - info: host.xxxxxx Restart adapter system.adapter.zigbee.0 because enabled
```

Right before I started my investigations, I'd updated the adapter to the most recent version `1.2.1`.

**Comment by tsvsj** [comment]
Additionally: Is there any description for the created objects? I don't understand the difference between `heating_setpnt_current`, `heating_setpnt_occupied` and `heating_setpnt_unoccupied`. Other objects aren't clear as well...

**Comment by asgothian** [comment]
Can you please provide a bit more information ? I have 3 of those thermostats in use, and none of the issues with the version i have installed (also 1.2.1)
- Which exact version 1.2.1 do you have installed (Github, npm?)
- Can you recreate the error with debug activated for the adapter, and post the last 10 messages from the zigbee adapter before the error occurs ?

Now for the data points. The Thermostat cluster supports different profiles to be defined wether the room is marked "occupied" or "unoccupied" (empty). As such, the Eurotronic thermostat supports these as well. I made these available as a firmware update might make these usable. The default cluster definition does NOT include a datapoint for the current target temperature - heating_setpnt_current is mapped to a manufacturer specific datapoint which may only be available in these Thermostats. Internally, both (heating_setpnt_occupied and heating_setpnt_unoccupied) are copied into heating_setpnt_current (as per the manual for the thermostat)

The Datapoint heating_demand reports the current valve position
The Datapoint local_temp reports the temperature measured by the thermostat (with the local_temp_calibration already taken into account). The combination of this value and the heating_setpnt_current value will control valve when in automatic mode.
The Datapoint spz_system_mode is a bitcoded combination of the datapoint 
- boost
- child_protection (Lock of control buttons)
- window_open
- mirror_display

The Datapoint spz_trv_mode allows to switch between 3 modes:
0 - OFF.
1 - Manual mode. Local temperature is reported, temperature calibration is used, but the valve is controlled using the valve_position datapoint
2 - Automatic mode. The Thermostat controls the valve based on heating_setpnt_current and its internal regulator. Setting heating_demand in this mode is not supported.

The Datapoint local_temp_calibration is an offset which is added to the temperature measured by the thermostat before it is reported and/or used for valve control. Min/Max value for this is +15/-15 degrees

I hope this helps.

**Comment by tsvsj** [comment]
Thanks for your explanation. That helps a lot. Do you see a chance to put this into the wiki or somethin similar?

> * Which exact version 1.2.1 do you have installed (Github, npm?)
NPM. I've just used the standard mechanism of ioBroker to update the adapter, but working with latest, instead of stable.

> * Can you recreate the error with debug activated for the adapter, and post the last 10 messages from the zigbee adapter before the error occurs ?
I hope, that all sensitive data is removed completely:
```
2020-10-06 08:49:51.938  - debug: zigbee.0 (5684) User stateChange zigbee.0.<XXXXXXXXXXXXXXX>.spz_trv_mode {"val":2,"ack":false,"ts":1601966991935,"q":0,"from":"system.adapter.admin.0","user":"system.user.admin","lc":1601966991935}
2020-10-06 08:49:51.945  - debug: zigbee.0 (5684) Change state 'spz_trv_mode' at device 0x<XXXXXXXXXXXXXXX> type 'SPZB0001'
2020-10-06 08:49:51.948  - debug: zigbee.0 (5684) State changes. dev: 0x<XXXXXXXXXXXXXXX> model: SPZB0001 states: [{"stateDesc":{"id":"spz_trv_mode","name":"TRV Mode","prop":"eurotronic_trv_mode","role":"state","write":true,"read":true,"type":"number"},"value":2,"index":0,"timeout":0}] opt: {"window_open":false,"boost":false,"child_protection":false,"mirror_display":false}
2020-10-06 08:49:51.950  - debug: zigbee.0 (5684) entity: {"type":"device","device":{"ID":10,"_type":"EndDevice","_ieeeAddr":"0x<XXXXXXXXXXXXXXX>","_networkAddress":6933,"_manufacturerID":4151,"_endpoints":[{"ID":1,"profileID":260,"deviceID":769,"inputClusters":[0,1,3,513,25,10],"outputClusters":[0,1,3,4,513,25,10],"deviceNetworkAddress":6933,"deviceIeeeAddress":"0x<XXXXXXXXXXXXXXX>","clusters":{"genBasic":{"attributes":{"modelId":"SPZB0001","manufacturerName":"Eurotronic","powerSource":3,"zclVersion":2,"appVersion":22,"stackVersion":5,"hwVersion":35,"dateCode":"20191014","swBuildId":"22190930"}},"hvacThermostat":{"attributes":{"16387":2000,"16392":1,"localTemp":2050,"pIHeatingDemand":68,"occupiedHeatingSetpoint":2200,"unoccupiedHeatingSetpoint":1600}}},"_binds":[{"cluster":1,"type":"endpoint","deviceIeeeAddress":"<XXXXXXXXXXXXXX1>","endpointID":1},{"cluster":513,"type":"endpoint","deviceIeeeAddress":"<XXXXXXXXXXXXXX1>","endpointID":1}]}],"_manufacturerName":"Eurotronic","_powerSource":"Battery","_modelID":"SPZB0001","_applicationVersion":22,"_stackVersion":5,"_zclVersion":2,"_hardwareVersion":35,"_dateCode":"20191014","_softwareBuildID":"22190930","_interviewCompleted":true,"_interviewing":false,"meta":{"configured":3},"_lastSeen":1601966952450},"mapped":{"zigbeeModel":["SPZB0001"],"model":"SPZB0001","vendor":"Eurotronic","description":"Spirit Zigbee wireless heater thermostat","supports":"temperature, heating system control","fromZigbee":[{"cluster":"hvacThermostat","type":["attributeReport","readResponse"]},{"cluster":"genPowerCfg","type":["attributeReport","readResponse"]}],"toZigbee":[{"key":["occupied_heating_setpoint"]},{"key":["unoccupied_heating_setpoint"]},{"key":["local_temperature_calibration"]},{"key":["system_mode"]},{"key":["eurotronic_host_flags","eurotronic_system_mode"]},{"key":["eurotronic_error_status"]},{"key":["setpoint_raise_lower"]},{"key":["control_sequence_of_operation"]},{"key":["remote_sensing"]},{"key":["current_heating_setpoint"]},{"key":["eurotronic_trv_mode"]},{"key":["eurotronic_valve_position"]},{"key":["local_temperature"]}],"meta":{"configureKey":3}},"endpoint":"[Circular]","endpoints":"[Circular]","name":"0x<XXXXXXXXXXXXXXX>"}
2020-10-06 08:49:51.952  - debug: zigbee.0 (5684) convert eurotronic_trv_mode, 2, {}
2020-10-06 08:49:51.953  - debug: zigbee.0 (5684) target: {"ID":1,"profileID":260,"deviceID":769,"inputClusters":[0,1,3,513,25,10],"outputClusters":[0,1,3,4,513,25,10],"deviceNetworkAddress":6933,"deviceIeeeAddress":"0x<XXXXXXXXXXXXXXX>","clusters":{"genBasic":{"attributes":{"modelId":"SPZB0001","manufacturerName":"Eurotronic","powerSource":3,"zclVersion":2,"appVersion":22,"stackVersion":5,"hwVersion":35,"dateCode":"20191014","swBuildId":"22190930"}},"hvacThermostat":{"attributes":{"16387":2000,"16392":1,"localTemp":2050,"pIHeatingDemand":68,"occupiedHeatingSetpoint":2200,"unoccupiedHeatingSetpoint":1600}}},"_binds":[{"cluster":1,"type":"endpoint","deviceIeeeAddress":"<XXXXXXXXXXXXXX1>","endpointID":1},{"cluster":513,"type":"endpoint","deviceIeeeAddress":"<XXXXXXXXXXXXXX1>","endpointID":1}]}
2020-10-06 08:49:53.386  - debug: zigbee.0 (5684) handleMessage. {"type":"attributeReport","device":{"ID":5,"_type":"EndDevice","_ieeeAddr":"<XXXXXXXXXXXXXX2>","_networkAddress":60831,"_manufacturerID":4151,"_endpoints":[{"ID":1,"profileID":260,"deviceID":24321,"inputClusters":[0,3,25,18],"outputClusters":[0,4,3,5,25,18],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genBasic":{"attributes":{"65281":{"1":3035,"3":21,"4":17320,"5":88,"6":[0,1],"10":0,"151":0,"152":62587,"153":55900,"154":4},"modelId":"lumi.sensor_cube.aqgl01","manufacturerName":"LUMI","powerSource":3,"zclVersion":1,"appVersion":5,"stackVersion":2,"hwVersion":1,"dateCode":"20160704","swBuildId":"3000-0001"}}},"_binds":[]},{"ID":2,"profileID":260,"deviceID":24322,"inputClusters":[3,18],"outputClusters":[4,3,5,18],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genMultistateInput":{"attributes":{"presentValue":2}}},"_binds":[]},{"ID":3,"profileID":260,"deviceID":24323,"inputClusters":[3,12],"outputClusters":[4,3,5,12],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genAnalogInput":{"attributes":{"65285":500,"presentValue":14.040000915527344}}},"_binds":[]}],"_manufacturerName":"LUMI","_powerSource":"Battery","_modelID":"lumi.sensor_cube","_applicationVersion":5,"_stackVersion":2,"_zclVersion":1,"_hardwareVersion":1,"_dateCode":"20160704","_softwareBuildID":"3000-0001","_interviewCompleted":true,"_interviewing":false,"meta":{},"_lastSeen":1601966993383},"endpoint":"[Circular]","data":{"65281":"[Circular]"},"linkquality":99,"groupID":0,"cluster":"genBasic","meta":{"zclTransactionSequenceNumber":204,"manufacturerCode":4447,"frameControl":{"frameType":0,"manufacturerSpecific":true,"direction":1,"disableDefaultResponse":true}}}
2020-10-06 08:49:53.389  - debug: zigbee.0 (5684) Received Zigbee message from '<XXXXXXXXXXXXXX2>', type 'attributeReport', cluster 'genBasic', data '{"65281":{"1":3035,"3":21,"4":17320,"5":88,"6":[0,1],"10":0,"151":0,"152":62587,"153":55900,"154":4}}' from endpoint 1 with groupID 0
2020-10-06 08:49:53.392  - debug: zigbee.0 (5684) Type attributeReport device {"type":"device","device":{"ID":5,"_type":"EndDevice","_ieeeAddr":"<XXXXXXXXXXXXXX2>","_networkAddress":60831,"_manufacturerID":4151,"_endpoints":[{"ID":1,"profileID":260,"deviceID":24321,"inputClusters":[0,3,25,18],"outputClusters":[0,4,3,5,25,18],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genBasic":{"attributes":{"65281":{"1":3035,"3":21,"4":17320,"5":88,"6":[0,1],"10":0,"151":0,"152":62587,"153":55900,"154":4},"modelId":"lumi.sensor_cube.aqgl01","manufacturerName":"LUMI","powerSource":3,"zclVersion":1,"appVersion":5,"stackVersion":2,"hwVersion":1,"dateCode":"20160704","swBuildId":"3000-0001"}}},"_binds":[]},{"ID":2,"profileID":260,"deviceID":24322,"inputClusters":[3,18],"outputClusters":[4,3,5,18],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genMultistateInput":{"attributes":{"presentValue":2}}},"_binds":[]},{"ID":3,"profileID":260,"deviceID":24323,"inputClusters":[3,12],"outputClusters":[4,3,5,12],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genAnalogInput":{"attributes":{"65285":500,"presentValue":14.040000915527344}}},"_binds":[]}],"_manufacturerName":"LUMI","_powerSource":"Battery","_modelID":"lumi.sensor_cube","_applicationVersion":5,"_stackVersion":2,"_zclVersion":1,"_hardwareVersion":1,"_dateCode":"20160704","_softwareBuildID":"3000-0001","_interviewCompleted":true,"_interviewing":false,"meta":{},"_lastSeen":1601966993383},"mapped":{"zigbeeModel":["lumi.sensor_cube","lumi.sensor_cube.aqgl01"],"model":"MFKZQ01LM","vendor":"Xiaomi","description":"Mi/Aqara smart home cube","supports":"shake, wakeup, fall, tap, slide, flip180, flip90, rotate_left and rotate_right","fromZigbee":[{"cluster":"genBasic","type":["attributeReport","readResponse"]},{"cluster":"genMultistateInput","type":["attributeReport","readResponse"]},{"cluster":"genAnalogInput","type":["attributeReport","readResponse"]}],"toZigbee":[]},"name":"<XXXXXXXXXXXXXX2>"} incoming event: {"type":"attributeReport","device":{"ID":5,"_type":"EndDevice","_ieeeAddr":"<XXXXXXXXXXXXXX2>","_networkAddress":60831,"_manufacturerID":4151,"_endpoints":[{"ID":1,"profileID":260,"deviceID":24321,"inputClusters":[0,3,25,18],"outputClusters":[0,4,3,5,25,18],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genBasic":{"attributes":{"65281":{"1":3035,"3":21,"4":17320,"5":88,"6":[0,1],"10":0,"151":0,"152":62587,"153":55900,"154":4},"modelId":"lumi.sensor_cube.aqgl01","manufacturerName":"LUMI","powerSource":3,"zclVersion":1,"appVersion":5,"stackVersion":2,"hwVersion":1,"dateCode":"20160704","swBuildId":"3000-0001"}}},"_binds":[]},{"ID":2,"profileID":260,"deviceID":24322,"inputClusters":[3,18],"outputClusters":[4,3,5,18],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genMultistateInput":{"attributes":{"presentValue":2}}},"_binds":[]},{"ID":3,"profileID":260,"deviceID":24323,"inputClusters":[3,12],"outputClusters":[4,3,5,12],"deviceNetworkAddress":60831,"deviceIeeeAddress":"<XXXXXXXXXXXXXX2>","clusters":{"genAnalogInput":{"attributes":{"65285":500,"presentValue":14.040000915527344}}},"_binds":[]}],"_manufacturerName":"LUMI","_powerSource":"Battery","_modelID":"lumi.sensor_cube","_applicationVersion":5,"_stackVersion":2,"_zclVersion":1,"_hardwareVersion":1,"_dateCode":"20160704","_softwareBuildID":"3000-0001","_interviewCompleted":true,"_interviewing":false,"meta":{},"_lastSeen":1601966993383},"endpoint":"[Circular]","data":{"65281":"[Circular]"},"linkquality":99,"groupID":0,"cluster":"genBasic","meta":{"zclTransactionSequenceNumber":204,"manufacturerCode":4447,"frameControl":{"frameType":0,"manufacturerSpecific":true,"direction":1,"disableDefaultResponse":true}}}
2020-10-06 08:49:53.397  - debug: zigbee.0 (5684) Publish {"battery":100,"voltage":3035}
2020-10-06 08:50:01.964  - error: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch().
2020-10-06 08:50:01.965  - error: zigbee.0 (5684) unhandled promise rejection: Timeout - 6933 - 1 - 2 - 513 - 4 after 10000ms
2020-10-06 08:50:01.966  - error: zigbee.0 (5684) Error: Timeout - 6933 - 1 - 2 - 513 - 4 after 10000ms
    at Timeout._onTimeout (/opt/iobroker/node_modules/zigbee-herdsman/dist/utils/waitress.js:44:24)
    at listOnTimeout (internal/timers.js:549:17)
    at processTimers (internal/timers.js:492:7)
2020-10-06 08:50:01.968  - info: zigbee.0 (5684) cleaned everything up...
2020-10-06 08:50:01.972  - info: zigbee.0 (5684) Zigbee: disabling joining new devices.
...
2020-10-06 08:50:01.978  - debug: zigbee.0 (5684) Publish available for 0x<XXXXXXXXXXXXXXX> = false
2020-10-06 08:50:02.254  - info: zigbee.0 (5684) terminating
2020-10-06 08:50:02.256  - info: zigbee.0 (5684) Terminated (NO_ERROR): Without reason
2020-10-06 08:50:03.245  - info: host.XXXXXX instance system.adapter.zigbee.0 terminated with code 0 (NO_ERROR)
2020-10-06 08:50:03.246  - info: host.XXXXXX Restart adapter system.adapter.zigbee.0 because enabled
```
Before the logs were created I've tried to set `spz_trv_mode` to value `2`. The value was unset before.

**Comment by asgothian** [comment]
I have to look at it. I can change the settings without this happening. Can you please post
- the version of the zigbee-herdsman-converters
- the version of the zigbee-herdsman
- the kind of Zigbee stick you use
- the firmware version on that stick
- which OS you are using
- Node and npm version

To check the version of any of the library components identify the package.json file within the respective libraries folder. The libraries can be found either in \opt\iobroker\node_modules or \opt\iobroker\node_modules\iobroker.zigbee\node_modules

Whichever datapoint i attempt to change the adapter stays stable.

You can try to install my debug version from https://github.com/asgothian/ioBroker.zigbee

**Comment by tsvsj** [comment]
The thermostat is the first device, that I try to control via ioBroker. All existing devices are just sensors or switches. All of them send data, instead of receiving data on a specific event (i.e. setting a value). I'm afraid that this may be just an issue related to the signal strength. The log line
```
2020-10-06 08:50:01.966  - error: zigbee.0 (5684) Error: Timeout - 6933 - 1 - 2 - 513 - 4 after 10000ms
```
shows definitely a timeout issue, because the node doesn't respond within the expected timeframe of 10secs. It seems, that this is an unhandled exception in the code, that may be considered as bug.

However, to circumvent deep investigations, because of a signal issue, I've ordered a smart plug, in order to extend the signal quality between the coordinator and node. The hardware should arrive by tomorrow. I will update asap with my results.

I'm using the probably most common setup, with a CC2531 USB Stick (w/o external antenna) and Z-stack firmware.

**Comment by asgothian** [comment]
The issue is the crash, not the timeout. If it was just a timeout, things would be fine, but the adapter should not crash when it cannot send a message.

**Comment by tsvsj** [comment]
The issue is found and solved. While integrating the innr SP220 I've recognized, that the version of the adapter was not 1.2.1, but 1.1.1. I'm running a multihost environment, where the CC2531 is only available in a slave node. It's quite confusing, that I can't update adapters from the master node, which reside only on slave nodes. The Zigbee adapter is only running on the slave node.

I'm sorry for that. I've updated the adapter to 1.3.0 and now everything works correctly. No crashes while changing a value. Since the smart plug works as a router the signal strength is at 80%. I'm sure the issue was related to the specific version 1.1.1.

Thanks for your help! I'm closing this issue.
